main/journal_to_versioned_journal.es6.js

```diff
@@ -0,0 +1,246 @@
+datastoreCtx.webInterfaceDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedWebInterface.id, pkg.VersionedWebInterface, ctx);
+datastoreCtx.releaseWebInterfaceJunctionDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedReleaseWebInterfaceJunction.id,
+    pkg.VersionedReleaseWebInterfaceJunction, ctx);
+datastoreCtx.browserMetricsDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedBrowserMetricData.id, pkg.VersionedBrowserMetricData, ctx);
+datastoreCtx.apiVelocityDAO.delegate = getOverwriteJournalDAO(
+    pkg.VersionedApiVelocityData.id, pkg.VersionedApiVelocityData, ctx);
+
+// SyncDAOs connected to Datastore.
+const releaseSyncDAO = datastoreCtx.releaseDAO;
+const webInterfaceSyncDAO = datastoreCtx.webInterfaceDAO;
+const releaseWebInterfaceJunctionSyncDAO =
+    datastoreCtx.releaseWebInterfaceJunctionDAO;
+const browserMetricsSyncDAO = datastoreCtx.browserMetricsDAO;
+const apiVelocitySyncDAO = datastoreCtx.apiVelocityDAO;
+
+// An unversioned cache of current Datastore.
+const releaseCacheDAO = cacheCtx.releaseDAO;
+const webInterfaceCacheDAO = cacheCtx.webInterfaceDAO;
+const releaseWebInterfaceJunctionCacheDAO =
+    cacheCtx.releaseWebInterfaceJunctionDAO;
+const browserMetricsCacheDAO = cacheCtx.browserMetricsDAO;
+const apiVelocityCacheDAO = cacheCtx.apiVelocityDAO;
+
+// In-memory DAOs of new data to be imported.
+const releaseImportDAO = importCtx.releaseDAO;
+const webInterfaceImportDAO = importCtx.webInterfaceDAO;
+const releaseWebInterfaceJunctionImportDAO =
+    importCtx.releaseWebInterfaceJunctionDAO;
+const browserMetricsImportDAO = importCtx.browserMetricsDAO;
+const apiVelocityImportDAO = importCtx.apiVelocityDAO;
+
+// Persistent DAOs of data that is exported. These will be later decorated with
+// versioning after Datastore versions are synced.
+const releaseExportDAO = datastoreCtx.releaseDAO.delegate;
+const webInterfaceExportDAO = datastoreCtx.webInterfaceDAO.delegate;
+const releaseWebInterfaceJunctionExportDAO =
+    datastoreCtx.releaseWebInterfaceJunctionDAO.delegate;
+const browserMetricsExportDAO = datastoreCtx.browserMetricsDAO.delegate;
+const apiVelocityExportDAO = datastoreCtx.apiVelocityDAO.delegate;
+
+//
+// Generic algorithm for data import:
+// (1) In parallel:
+// (1a) Sync and unversion Datastore data;
+// (1b) Load/compute new data;
+// (2) Import new data by put()ing what's changed and remove()ing anything in
+//     unversioned Datastore cache that does not also appear in new data.
+//
+
+const updater = pkg.DatastoreUpdater.create();
+function doImport(sync, load, daosArray) {
+  return Promise.all([
+    Promise.all(daosArray.map(sync)).then(() => {
+      return Promise.all(daosArray.map(daos => {
+        return updater.unversionData(daos.sync, daos.cache);
+      }));
+    }),
+    load(),
+  ]).then(() => {
+    return Promise.all(daosArray.map(function(daos) {
+      return updater.importData(daos.import, daos.cache, daos.export);
```

Comment:
I think this only writes the delta between cache and import. The export DAO was truncated, so it will contain only the delta, not the whole set.

Reply:
Yes and no. The journals are truncated. However, they are installed as the delegate to the Datastore `SyncDAO` instances. As a result, waiting for `theSyncDAO.sync()` pushes the data currently in Datastore into the journal first, then `updater.importData()` appends the diff to the journal.

Unrelated: This does mean that data that are changed (rather than newly added) appear in the journal twice, and also means that we don't have a journal that contains *just* the changes that should later be pushed to Datastore.

Reply:
Also, I extended the comment in the export DAO initialization to clarify how sync-then-add-updates works.